### ISSUE TITLE
Please add wxPerl

### ISSUE THREAD

**Issue by sciurius** [StrawberryPerl/build-extlibs#62]
There is a growing interest for wxPerl based applications. Would it be possible to add wxWidgets 3.2 and wxPerl to the Strawberry kit?

Even though the official wxPerl development has stalled for several years, modern Linux distributions provide wxWidgets 3.2 and wxPerl, so most porting problems have already been solved.

**Comment by ChordPro** [comment]
The [ChordPro application](https://chordpro.org/) is written in wxPerl and available for Linux, Windows and macOS. According to the downloads from GitHub and the [user forum](https://groups.io/g/ChordPro) we estimate that there are a couple of thousands ChordPro users on Windows.

For Windows we needed custom crafted packaging so we could include Strawberry Perl, wxWidgets 3.2 and (an updated version of) wxPerl.

For our development, and for the benefit of Windows users that know how to handle software installs it would be very helpful if wxWidgets and wxPerl were included in the Strawberry kit.

**Comment by eekboek** [comment]
[EekBoek](https://www.eekboek.nl) is a bookkeeping program for dutch companies. It is written in wxPerl and included in several Linux distributions. There is also a Windows version.

The Windows version was produced using Citrus Perl but moved to Strawberry Perl with wxWidgets 3.0 a couple of years ago. This took substantial efforts and keeps us from upgrading to wxWidgets 3.2. Having wxWidgets and wxPerl in the Strawberry Perl kit would be great!

Of course, we are willing to help and share experiences.

**Comment by oodler577** [comment]
A million times yes for me. This would enable a huge population of Perl programmers to spread Perl to Windows and make a living by their own work. If this happened, I would be extremely likely to donate to this project.

**Comment by shawnlaffan** [comment]
It comes down to the maintenance burden.  Each package adds build and potentially debugging time for new releases.  

To quote from a similar request to package Tk https://github.com/StrawberryPerl/Perl-Dist-Strawberry/issues/87#issuecomment-1475078019 :
>  If a module can be installed easily then we can look at it. If it is more involved then we are open to pull requests, so long as the maintenance burden is low. If a module is difficult to install then that is often best taken up with the maintainers of the module.

Perhaps not an issue here, but we also don't want to greatly increase the size of the downloads.

One option is to look at developing an Alien module for Wx, noting that build options and patches can be derived from the MSYS2 project.  https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-wxwidgets3.2

**Comment by sisyphus** [comment]
> Perhaps not an issue here, but we also don't want to greatly increase the size of the downloads.

Y'know, that could be addressed by using PPM. (PPM is still packaged with Strawberry Perl.)
Just build the standard SP perl release (as per is currently done) and include, in that standard release, the ppd files for any add-ons that you're making available.
Then, users can install those extra packages by simply running <code>ppm install PDL.ppd</code> or <code>ppm install Tk.ppd</code>, etc.
It also enables the possibility of having these particular packages upgraded, without having to re-build perl.

I'm a bit puzzled that PDL is not already handled in that way. (I guess storage space on the SP website is not really much of an issue.)

Once the module has been built, the creation of the PPM package can usually be done quite quickly. Tk and PDL should be straightforward, though I've no experience of wxPerl.

However, there's no easy way around the extra work of building those additional modules in the first place ;-)

It has been a while, but I think I could send you a PPM package (ie tarball and ppd file) for Tk - if you're interested.
To create the ppd file, I would need to know the location (on the SP server) where the tarball will reside.
Another advantage of this approach is that the creation of the PPM packages can be farmed out to other volunteers. It doesn't all have to be done by Shawn Laffan.

**Comment by sciurius** [comment]
> If a module can be installed easily 

If wxPerl could be built and installed easily we would probably not be discussing this. But instead of several individuals each spending time and effort to get it going it is much better to have this centralized so we can share resources. StrawberryPerl/build-exlibs seemed to me a good central place.

Whether wxPerl and wxWidgets would be part of the core distribution or PPM packages is not crucial. PPM packages is fine. This also makes it possible to upgrade a package without depending on a Strawberry release.

More important is whether the Strawberry build-exlibs team is willing to rebuild the packages for a new Strawberry release, assuming that once we have this figured out rebuilding will be near trivial.

EDIT: To use full functionality a customized build of wxWidgets is required. We can't just pull in the prebuilt binaries from the wxWidgets site.

**Comment by oodler577** [comment]
> StrawberryPerl/build-exlibs seemed to me a good central place.

I agree. And I think the alternative central location is the MSYS2 repository. I have looked into building such a package, and it seems doable; but that also means we'd be using the mingw perl package, which is not Strawberry Perl. FWIW, it is trival to get wxPerl installed on Debian via apt because they have put in the work, focusing on their central place of distributing software.

To move to the realm of speculation (i.e., I can't prove any of this), but including wxPerl in Strawberry Perl would make this distribution of Perl (and Perl itself), an extremely attractive option of application development on Windows - e.g., it would instantly get a placeon wxWidget's site as a way to get things going on Windows if it was really just as easy as, "install Strawberry Perl". It would draw many, many people to Perl that may not have considered it in the past. This would make it more likely also that people would donate to the project for th efforts. Focusing on application development (and eventually distribution) of Perl-based GUI programs on Windows is a key part of securing Perl's future - not just on Windows, in general. This is why I am willing to assist within my areas of capabilities or even try things that would stretch that to help. Thank you for your consideration.

**Comment by sciurius** [comment]
For your convenience, I've forked `Alien::wxWidgets` and 'released' an [updated version](https://github.com/sciurius/perl-Alien-wxWidgets/releases) that behaves satisfactory with Strawberry Perl.

[Extensive instructions for Strawberry users](https://github.com/sciurius/perl-Alien-wxWidgets/blob/master/Strawberry.md).

Bottom line is that with the updated versions of `Alien::wxWidgets` and `Wx` it should now be easy to get wxPerl on Strawberry, without having to deal with wxWidgets building/installing.

**Comment by shawnlaffan** [comment]
> For your convenience, I've forked Alien::wxWidgets and 'released' an [updated version](https://github.com/sciurius/perl-Alien-wxWidgets/releases) that behaves satisfactory with Strawberry Perl.

Thanks for doing this.  

One suggestion I would make is to convert it to use [Alien::Build](https://metacpan.org/pod/Alien::Build) with an `alienfile`.  The `A::B` infrastructure can handle all the searching, downloading and building which means it can be installed using a plain `cpanm` call without needing to unpack a tarball.

**Comment by sciurius** [comment]
Thanks for the suggestion, but I do not see any advantage since the `Alien::wxWidgets` build already takes care of that. Right now you can use
````
cpanm https://github.com/sciurius/perl-Alien-wxWidgets/releases/download/R0.70/Alien-wxWidgets-0.71.tar.gz
````
and everything will be downloaded, built and installed.

The 'extensive instructions' are merely finetuning for cases where wxWidgets may already have been downloaded and installed.

**Comment by shawnlaffan** [comment]
> Thanks for the suggestion, but I do not see any advantage since the `Alien::wxWidgets` build already takes care of that.

Fair enough.  

I think this issue is now closable?  Any further discussions should probably occur on the [module's issue tracker](https://github.com/sciurius/perl-Alien-wxWidgets/issues).

**Comment by oodler577** [comment]
You can keep this closed, but I had a question. What reasoning justified the PDL distribution to be created?

**Comment by shawnlaffan** [comment]
> What reasoning justified the PDL distribution to be created?

That long pre-dates my involvement in the project.  It was first released with 5.20, in 2014.  https://strawberryperl.com/releases.html

Perhaps others can provide more specific details, but PDL did used to be quite difficult to build and install.

**Comment by sisyphus** [comment]
> What reasoning justified the PDL distribution to be created?

I think that would have been a decision made by @kmx (aka @karel-m ?). As @shawnlaffan noted, it wasn't a braindead-straightforward build back then.
It's inclusion is uncontroversial IMO. Is there a more important scientific perl module ?

**Comment by oodler577** [comment]
@sisyphus -

In the event that a wxPerl specific or "Windows GUI development" distribution of Strawberry Perl is warranted, it'd be helpful to know what reasoning was used to justify a PDL specific version. This has nothing to do with the Strawberry Perl that comes with PDL, it has to do with the reasons used to justify the effort.

**Comment by sisyphus** [comment]
@oodler577, IIRC, no justification for the PDL build was ever given. It just started being provided, and no-one has ever complained about that. One can take it or leave it.
I speculate that, although it entailed some extra work, kmx decided that he had enough personal interest in PDL to justify the extra load .... but that is just my speculation.

**Comment by oodler577** [comment]
> [@oodler577](https://github.com/oodler577), IIRC, no justification for the PDL build was ever given. It just started being provided, and no-one has ever complained about that. One can take it or leave it. I speculate that, although it entailed some extra work, kmx decided that he had enough personal interest in PDL to justify the extra load .... but that is just my speculation.

Understood, thank you! :-)

**Comment by adrian-vasile-constantin** [comment]
Many thanks to @sciurius for implementing this
